### PR TITLE
Reduce open goroutines

### PR DESCRIPTION
**Pull request by Gusted**
- When bleve isn't used as indexer, close the 4 open goroutines they have open by default.
- When bleve is being used, use the configured amount of analysis workers. By default 1, as I don't assume issues/repo's are being opened every ms, which roughly how long it takes in order to analyze by bleve. For instances where this is the case, they can increase it back to 4 or more, or move to Elasticsearch.

- Ref: https://github.com/blevesearch/bleve/blob/03c6595327f10412e7e116679691d0438b776cf6/config.go#L56 (This function is called in `init()` Line 67)

Before:
4 open goroutines
image

After(indexer is bleve):
1 open goroutine
image

### CODE REVIEW

File: modules/indexer/issues/indexer.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ import (
 	"code.gitea.io/gitea/modules/queue"
 	"code.gitea.io/gitea/modules/setting"
 	"code.gitea.io/gitea/modules/util"
+	"github.com/blevesearch/bleve/v2"
```
Comment (techknowlogick) [inline modules/indexer/issues/indexer.go:23]: Please add an empty line ahead of this one

Comment (wxiaoguang) [reply modules/indexer/issues/indexer.go:23]: Our own `codeformat` should do the work with `make fmt`, but now our own `codefomtat` isn't called. We should invite it back 😊

**Comment by techknowlogick** [comment]
@lunny I'm not sure this is breaking.

**Comment by lunny** [comment]
> @lunny I'm not sure this is breaking.

The default goroutines number reduced from 4 to 1 so that the default index speed should be slower than before.

**Comment by Gusted** [comment]
> The default goroutines number reduced from 4 to 1 so that the default index speed should be slower than before.

Not really, it only happens in specific cases. The only code where analysis is being done:

https://github.com/blevesearch/bleve/blob/03c6595327f10412e7e116679691d0438b776cf6/index/scorch/scorch.go#L365

From what I can see when the Index is being called with multiple index operations is for the code indexer & migrations(? I don't see any code for that, but I assume they index issues after migration is completed) (in which yes, the speed in slower), but by default it's only being called with single operations(indexing new issues).

File: modules/indexer/issues/indexer.go (hunk 1 above)
Comment (6543) [reply modules/indexer/issues/indexer.go:23]: on it .-..

File: modules/indexer/issues/indexer.go [hunk 2]
```diff
@@ -102,6 +104,13 @@ var (
 func InitIssueIndexer(syncReindex bool) {
 	waitChannel := make(chan time.Duration)
 
+	if setting.Indexer.IssueType != "bleve" {
+		// When bleve isn't used, shutdown any open goroutines they have.
+		bleve.Config.Shutdown()
```
Comment (6543) [inline modules/indexer/issues/indexer.go:109]: we did not do that atm - is bleve already initialized at this point ?!?

Comment (Gusted) [reply modules/indexer/issues/indexer.go:109]: Bleve is initialized by the init functions, so it is initialized at this point.

Comment (Gusted) [reply modules/indexer/issues/indexer.go:109]: See the ref in OP comment. That's the init function.

**Comment by wxiaoguang** [comment]
After reading the bleve code, I prefer to leave bleve as it was, instead of setting paramters/shutdown bleve by ourselves.

The reason is:
* bleve's init code is quite dirty. As long as we import the `bleve` package, the `init` will be called and start some go-routines. 
* go-routine is quite lightweight, 0 vs 1 vs 4 doesn't bring too much difference, we can just ignore them.

If we want to make the go-routines clear, we had better ask bleve to make their code clear: do not create go-routines in `init`

**Comment by Gusted** [comment]
> If we want to make the go-routines clear, we had better ask bleve to make their code clear: do not create go-routines in `init`

That would be a pretty big breaking change for them, but I would like to see a issue on their repo for this request, as this was undefined behavior for me and probably a lot of other developers that are using bleve that suddenly have goroutines hanging around. I will make a issue soon(if nobody else is sooner)

**Comment by Gusted** [comment]
Closing for now, as this is clearly not the way to go.

**Comment by wxiaoguang** [comment]
> > If we want to make the go-routines clear, we had better ask bleve to make their code clear: do not create go-routines in `init`
> 
> That would be a pretty big breaking change for them, but I would like to see a issue on their repo for this request, as this was undefined behavior for me and probably a lot of other developers that are using bleve that suddenly have goroutines hanging around. I will make a issue soon(if nobody else is sooner)

Agree to find a solution from upstream. For Gitea, it just uses a lot of 3rd libraries, it's not ideal to always patch to bypass the upstream problems.

Actually if there is a serious problem and no fix from upstream, then we have to patch it. But if upstream could fix the problem or the problem is not that serious, it's better to keep Gitea code clear, otherwise there would be a lot of patches which are difficult to maintain 😂.

**Pull request closed without merging**